lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -357,6 +345,90 @@ def from_string(private_key: str) -> "PrivateKey":
         return PrivateKey(raw=private_key.strip())
 
 
+@dataclass(frozen=True)
+class Certificate:
+    """This class represents a certificate."""
+
+    raw: str
+    common_name: str
+    sans_dns: Optional[List[str]] = None
+    sans_ip: Optional[List[str]] = None
+    sans_oid: Optional[List[str]] = None
+    email_address: Optional[str] = None
+    organization: Optional[str] = None
+    organizational_unit: Optional[str] = None
+    country_name: Optional[str] = None
+    state_or_province_name: Optional[str] = None
+    locality_name: Optional[str] = None
+    expiry_time: Optional[datetime] = None
+    validity_start_time: Optional[datetime] = None
+
+    def __str__(self) -> str:
+        """Return the certificate as a string."""
+        return self.raw
+
+    @staticmethod
+    def from_string(certificate: str) -> "Certificate":
+        """Create a Certificate object from a certificate."""
+        try:
+            certificate_object = x509.load_pem_x509_certificate(data=certificate.encode())
+        except ValueError as e:
+            logger.error("Could not load certificate: %s", e)
+            raise TLSCertificatesError("Could not load certificate")
+        common_name = certificate_object.subject.get_attributes_for_oid(NameOID.COMMON_NAME)
+        country_name = certificate_object.subject.get_attributes_for_oid(NameOID.COUNTRY_NAME)
+        state_or_province_name = certificate_object.subject.get_attributes_for_oid(
+            NameOID.STATE_OR_PROVINCE_NAME
+        )
+        locality_name = certificate_object.subject.get_attributes_for_oid(NameOID.LOCALITY_NAME)
+        organization_name = certificate_object.subject.get_attributes_for_oid(
+            NameOID.ORGANIZATION_NAME
+        )
+        email_address = certificate_object.subject.get_attributes_for_oid(NameOID.EMAIL_ADDRESS)
+        try:
+            sans = certificate_object.extensions.get_extension_for_class(
+                x509.SubjectAlternativeName
+            ).value
+            sans_dns = [
+                str(san)
+                for san in sans.get_values_for_type(x509.DNSName)
+                if isinstance(san, x509.DNSName)
+            ]
+            sans_ip = [
+                str(san)
+                for san in sans.get_values_for_type(x509.IPAddress)
+                if isinstance(san, x509.IPAddress)
+            ]
+            sans_oid = [
+                str(san)
+                for san in sans.get_values_for_type(x509.RegisteredID)
+                if isinstance(san, x509.RegisteredID)
+            ]
+        except x509.ExtensionNotFound:
+            sans_dns = []
+            sans_ip = []
+            sans_oid = []
+        expiry_time = certificate_object.not_valid_after_utc
+        validity_start_time = certificate_object.not_valid_before_utc
+
+        return Certificate(
```

Comment:
```suggestion
        return cls(
```

And here, we can use the `cls` parameter instead, that would let us subclass this in the future.

Reply:
I'll resolve this one here so that we keep the discussion in 1 place.